ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/VolDB.hs

```diff
@@ -268,47 +273,221 @@ candidates succsOf b = mapMaybe NE.nonEmpty $ go (fromChainHash (pointHash b))
+--
+-- When successful, return a 'PointChainDiff': the number of blocks to roll
+-- back the chain fragment to the intersection point, an anchor corresponding
+-- to that intersection, and a list of points corresponding to the
+-- blocks/headers needed to connect the intersection point to @P@. The
+-- intersection point will be the most recent intersection point.
+--
+-- Returns 'Nothing' when @P@ is not in the VolatileDB or when @P@ is not
+-- connected to the given chain fragment.
 --
--- Returns a path (of hashes) from @I@ (excluded) to @B@ (included).
+-- POSTCONDITION: the returned number of blocks to roll back is less than or
+-- equal to the length of the given chain fragment.
 --
--- PRECONDITION: @B ∈ V@, i.e. @B@ is part of the VolatileDB.
+-- Note that the number of returned points can be smaller than the number of
+-- blocks to roll back. This means @P@ is on a fork shorter than the given
+-- chain fragment.
 --
--- 'True' <=> for all transitive predecessors @B'@ of @B@ we have @B' ∈ V@ or
--- @B' = I@.
-isReachable :: forall blk. (HasHeader blk, HasCallStack)
-            => (HeaderHash blk -> Maybe (WithOrigin (HeaderHash blk)))
-            -> Point blk                                        -- ^ @I@
-            -> RealPoint blk                                    -- ^ @B@
-            -> Maybe (NonEmpty (HeaderHash blk))
-isReachable predecessor i b =
-    case computePath predecessor from to of
-      -- Bounds are not on the same fork
-      Nothing   -> Nothing
-      Just path -> case path of
-        CompletelyInVolDB hashes -> case NE.nonEmpty hashes of
-          Just hashes' -> Just hashes'
-          Nothing      ->
-            error "impossible: empty list of hashes with an inclusive bound"
-        PartiallyInVolDB  {}     -> Nothing
-        NotInVolDB        _      ->
-          error "impossible: block just added to VolatileDB missing"
+-- A 'PointChainDiff' is returned iff @P@ is on the chain fragment. Moreover,
+-- when the number of blocks to roll back is also 0, it must be that @P@ is
+-- the tip of the chain fragment.
+--
+-- When the returned list of points is non-empty, @P@ will be the last point
+-- in the suffix of the 'PointChainDiff'.
+isReachable
+  :: forall blk. ((HasHeader (Header blk)), HasCallStack)
+  => LookupBlockInfo blk
+  -> AnchoredFragment (Header blk)
+  -> RealPoint blk
+  -> Maybe (PointChainDiff blk)
+isReachable lookupBlockInfo = \chain b ->
+    case computeReversePath lookupBlockInfo b of
+      -- Block not in the VolatileDB, so it's unreachable
+      Nothing          -> Nothing
+      Just reversePath -> go chain reversePath 0 []
   where
-    from = StreamFromExclusive i
-    to   = StreamToInclusive   b
+    -- | NOTE: the 'ReversePath' is lazy in its spine. We will only force as
+    -- many elements as 'RealPoint's we return. In the worst case, the path is
+    -- not connected to the current chain at all, in which case we do force
+    -- the entire path.
+    go
```

Comment:
Document: we're trying to find a common block, i.e., one with the same point and thus the same slot. Both the chain and the path are ordered by slots, so we compare the slots and drop the largest one until we have a match in slot, then we check hashes. If those don't match, we drop both.